File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def check(l) :
    t = []
    l = list(l)
    for i in l :
        if not t:
            t.append(i)
        else :
            if (i == ')' and t[-1] =='(') or (i == ']' and t[-1] == '[') or (i == '}' and t[-1] == '{') :
                t.pop()
            else :
                t.append(i)
    if t : return False
    else : return True
            
from collections import deque            
def solution(s):
    answer = 0
    queue = deque(s)
    for i in range(len(s)) :
        queue.rotate(-i)
        if check(queue) :
            answer += 1
        queue.rotate(i)
    return answer

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public static int solution(String s) {
        int answer = 0;
        List<Character> list = new ArrayList<>();
        char[] charArray = s.toCharArray();
        for (char c : charArray) {
            list.add(c);
        }

        for (int i = 0; i < list.size(); i++) {
            if (i != 0) {
                Collections.rotate(list, -1);
            }
            char[] chars = new char[list.size()];
            for (int j = 0; j < list.size(); j++) {
                chars[j] = list.get(j);
            }
            if (check(chars)) {
                answer++;
            }
        }
        return answer;
    }

    public static boolean check(char[] chars) {
        ArrayDeque<Character> stack = new ArrayDeque<>();

        for (char c : chars) {
            if (c == '(' || c == '[' || c == '{') {
                stack.addLast(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                } else if (c == ')' && stack.peekLast() == '(') {
                    stack.pollLast();
                } else if (c == ']' && stack.peekLast() == '[') {
                    stack.pollLast();
                } else if (c == '}' && stack.peekLast() == '{') {
                    stack.pollLast();
                }
            }
        }
        return stack.isEmpty();
    }
}